gwcs/wcs.py

```diff
@@ -634,3 +641,235 @@ def fix_inputs(self, fixed):
+            if ((max_error is not None and max_resid < max_error and max_rms is None)
+                 or (max_error is not None and max_resid < max_error and max_rms is not None
+                     and rms < max_rms)
+                 or (max_error is None and max_rms is not None and rms > max_rms)):
+                cond = False
+                if verbose:
+                    print('terminating condition met')
+                break
+            else:
+                degree += 1
+                if degree == 10:
+                    raise RuntimeError("Fit conditions not met")
+    return fit_poly_x, fit_poly_y, max_resid, rms
+
+
+def compute_distance_residual(undist_x, undist_y, fit_poly_x, fit_poly_y):
+    """
+    Compute the distance residuals and return the rms and maximum values.
+    """
+    dist = np.sqrt((undist_x - fit_poly_x)**2 + (undist_y - fit_poly_y)**2)
+    max_resid = dist.max()
+    rms = np.sqrt((dist**2).sum() / dist.size)
+    return max_resid, rms
+
+
+def reform_poly_coefficients(fit_poly_x, fit_poly_y):
+    """
+    The fit polynomials must be recombined to align with the SIP decomposition
+
+    The result is the f(u,v) and g(u,v) polynomials, and the CD matrix.
+    """
+    # Extract values for CD matrix and recombing
+    c11 = fit_poly_x.c1_0.value
+    c12 = fit_poly_x.c0_1.value
+    c21 = fit_poly_y.c1_0.value
+    c22 = fit_poly_y.c0_1.value
+    sip_poly_x = fit_poly_x.copy()
+    sip_poly_y = fit_poly_y.copy()
+    # Force low order coefficients to be 0 as defined in SIP
+    sip_poly_x.c0_0 = 0
+    sip_poly_y.c0_0 = 0
+    sip_poly_x.c1_0 = 0
+    sip_poly_x.c0_1 = 0
+    sip_poly_y.c1_0 = 0
+    sip_poly_y.c0_1 = 0
+    det = c11 * c22 - c12 * c21
+    ff = c22 / det
+    fg = c12 / det
+    gg = c11 / det
+    gf = c21 / det
+    degree = fit_poly_x.degree
+    # Now loop through all remaining coefficients
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j > 1) and (i + j < degree + 1):
+                old_x = getattr(fit_poly_x, f'c{i}_{j}').value
+                old_y = getattr(fit_poly_y, f'c{i}_{j}').value
+                new_x = ff * old_x - fg * old_y
+                setattr(sip_poly_x, f'c{i}_{j}', new_x)
+                new_y = gg * old_y - gf * old_x
+                setattr(sip_poly_y, f'c{i}_{j}', new_y)
+    cd = ((c11, c12), (c21, c22))
+    return cd, sip_poly_x, sip_poly_y
+
+
+def store_2D_coefficients(hdr, poly_model, coeff_prefix, keeplinear=False):
+    """
+    Write the polynomial model coefficients to the header.
+    """
+    if keeplinear:
+        mindeg = 0
+    else:
+        mindeg = 1
+    degree = poly_model.degree
+    for i in range(0, degree + 1):
+        for j in range(0, degree + 1):
+            if (i + j) > mindeg and (i + j < degree + 1):
+                hdr[f'{coeff_prefix}_{i}_{j}'] = getattr(poly_model, f'c{i}_{j}').value
```

Comment:
I think if `value` is identical to 0, the coefficient should not be written out.
```suggestion
                val = getattr(poly_model, f'c{i}_{j}').value
                if val != 0:
                    hdr[f'{coeff_prefix}_{i}_{j}'] = val
```

Reply:
... at least that is how `astropy.wcs.WCS.to_header()` works.

Reply:
I tend to like it explicit, but if @nden agrees with you, I'll change it.

Reply:
:+1: for explicit